question2/odbcdemo.c

```diff
@@ -0,0 +1,288 @@
+                SQLHANDLE hdbc,  /* ODBC Connection Handle */
+                SQLHANDLE hstmt) /* ODBC SQL Handle */
+{
+    UCHAR sqlstate[10];
+    UCHAR errmsg[SQL_MAX_MESSAGE_LENGTH];
+    SQLINTEGER nativeerr;
+    SQLSMALLINT actualmsglen;
+    RETCODE rc;
+loop:
+    rc = SQLError(henv, hdbc, hstmt, (SQLCHAR *)sqlstate, &nativeerr,
+                  (SQLCHAR *)errmsg, SQL_MAX_MESSAGE_LENGTH - 1, &actualmsglen);
+    if (rc == SQL_ERROR) {
+        printf("SQLError failed!\n");
+        return;
+    }
+    if (rc != SQL_NO_DATA_FOUND) {
+        printf("SQLSTATE = %s\n", sqlstate);
+        printf("NATIVE ERROR = %d\n", nativeerr);
+        errmsg[actualmsglen] = '\0';
+        printf("MSG = %s\n\n", errmsg);
+        goto loop;
+    }
+}
+
+/*
+** function: EnvClose
+**
+** Purpose: Frees environment and connection handles.
+**
+** Arguments: henv _ environment handle
+** hdbc - connection to handle
+*/
+void EnvClose(SQLHANDLE henv, SQLHANDLE hdbc) {
+    SQLDisconnect(hdbc);
+    SQLFreeHandle(SQL_HANDLE_DBC, hdbc);
+    SQLFreeHandle(SQL_HANDLE_ENV, henv);
+}
+
+/*
+** function: fgets_wrapper
+**
+** Purpose: Handles newline for fgets from stdin.
+**
+** Arguments: buffer _ character array ptr
+** buflen - max characters
+*/
+char *fgets_wrapper(char *buffer, size_t buflen) {
+    if (fgets(buffer, buflen, stdin) != 0) {
+        buffer[strcspn(buffer, "\n")] = '\0';
+        return buffer;
+    }
+    return 0;
+}
+
+/*
+** Defines used by main program.
+*/
+#define PWD_LEN 32
+#define UID_LEN 32
+#define DSN_LEN 32
+#define USAGE_MSG1 "Usage: %s \n"
+
+/*
+** Program: odbc demo
+**
+** Purpose: ODBC Demo routine.
+*/
+
+typedef struct {
+    UCHAR charCol1[1024];
+    // SQLLEN length1 ;
+    long length1;
+    short length2;
+} DataInfoStruct;
+
+DataInfoStruct dataStruct[32];
+DataInfoStruct dataStruct2[32];
+DataInfoStruct dataStruct3[32];
+
+int main(int argc, char *argv[]) {
+    SQLHANDLE hdbc;
+    SQLHANDLE henv;
+    SQLHANDLE hstmt;
+    RETCODE rc;
+    UCHAR uid[UID_LEN];
+    UCHAR pwd[PWD_LEN];
+    UCHAR driver[DSN_LEN];
+    UCHAR ver[32];
+    SQLSMALLINT strLen;
+    int i;
+
+    uid[0] = 0;
+    pwd[0] = 0;
+    if (argc > 1) {
+        printf(USAGE_MSG1, argv[0]);
+        return (1);
+    }
+
+    printf("\nEnter the DSN : ");
+    fgets_wrapper((char *)driver, DSN_LEN);
+    printf("\nEnter the UID : ");
+    fgets_wrapper((char *)uid, UID_LEN);
+    printf("\nEnter the PWD : ");
+    fgets_wrapper((char *)pwd, PWD_LEN);
+
+    printf("%s: will connect to data source '%s' as user '%s/%s'.\n", argv[0],
+           driver, uid, pwd);
+
+    rc = SQLAllocHandle(SQL_HANDLE_ENV, SQL_NULL_HANDLE, &henv);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("Unable to allocate environment\n");
+        exit(255);
+    }
+
+    rc = SQLSetEnvAttr(henv, SQL_ATTR_ODBC_VERSION, (SQLPOINTER)SQL_OV_ODBC3,
+                       SQL_IS_INTEGER);
+
```

Comment:
we need to allocate the connection handle before calling SQLConnect
```suggestion
    rc = SQLAllocHandle(SQL_HANDLE_DBC, henv, &hdbc);
    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
        printf("Unable to allocate connection\n");
        exit(255);
    }
```

Reply:
in addition, we should free our environment handle here
```suggestion
    rc = SQLAllocHandle(SQL_HANDLE_DBC, henv, &hdbc);
    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
        printf("Unable to allocate connection\n");
        SQLFreeHandle(SQL_HANDLE_ENV, henv);
        exit(255);
    }
```